src/puya/awst_build/eb/arc4/abi_call.py

```diff
@@ -0,0 +1,407 @@
+            # TODO: what about mapping int to BigUInt?
+            return wtypes.uint64_wtype
+        case Literal(value=str()):
+            return wtypes.arc4_string_wtype
+        case Literal(value=bool()):
+            return wtypes.bool_wtype
+        case _:
+            raise CodeError("Invalid arg type", arg.source_location)
+
+
+def _create_abi_call_expr(
+    method_selector: str,
+    abi_arg_types: Sequence[wtypes.WType],
+    abi_return_type: wtypes.WType,
+    abi_args: Sequence[ExpressionBuilder | Literal],
+    transaction_kwargs: dict[str, ExpressionBuilder | Literal],
+    location: SourceLocation,
+) -> Expression:
+    abi_arg_exprs: list[Expression] = [
+        MethodConstant(
+            value=method_selector,
+            source_location=location,
+        )
+    ]
+    asset_exprs = list[Expression]()
+    account_exprs = list[Expression]()
+    application_exprs = list[Expression]()
+
+    def append_ref_arg(ref_list: list[Expression], arg_expr: Expression) -> None:
+        # asset refs start at 0, account and application start at 1
+        implicit_offset = 0 if ref_list is asset_exprs else 1
+        # TODO: what about references that are used more than once?
+        ref_index = len(ref_list)
+        ref_list.append(arg_expr)
+        abi_arg_exprs.append(
+            BytesConstant(
+                value=(ref_index + implicit_offset).to_bytes(length=1),
+                encoding=BytesEncoding.base16,
+                source_location=arg_expr.source_location,
+            )
+        )
+
+    for arg, wtype in zip(abi_args, abi_arg_types, strict=True):
+        arg_expr = expect_operand_wtype(arg, wtype)
+        match wtype:
+            case wtypes.ARC4Type():
+                abi_arg_exprs.append(arg_expr)
+            case wtypes.asset_wtype:
+                append_ref_arg(asset_exprs, arg_expr)
+            case wtypes.account_wtype:
+                append_ref_arg(account_exprs, arg_expr)
+            case wtypes.application_wtype:
+                append_ref_arg(application_exprs, arg_expr)
+            case _ if wtypes.has_arc4_equivalent_type(wtype):
+                arc4_wtype = wtypes.avm_to_arc4_equivalent_type(wtype)
+                abi_arg_exprs.append(arc4_encode(arg_expr, arc4_wtype, arg_expr.source_location))
+            case wtypes.WGroupTransaction():
+                raise CodeError(
+                    "Transaction arguments are not supported for contract to contract calls",
+                    arg_expr.source_location,
+                )
+            case _:
+                raise CodeError("Invalid argument type", arg_expr.source_location)
+
+    fields: dict[TxnField, Expression] = {
+        TxnFields.type: UInt64Constant(
+            value=constants.TransactionType.appl.value,
+            teal_alias=constants.TransactionType.appl.name,
+            source_location=location,
+        )
+    }
+    if abi_arg_exprs:
+        fields[TxnFields.app_args] = TupleExpression(
```

Comment:
Do we need to do any special handling of 15+ args?

Reply:
Yep, still need to implement that

Reply:
Done